ocw/lib/db.py

```diff
@@ -166,40 +191,32 @@ def update_run():
     global __running, __last_update
     __running = True
     cfg = ConfigFile()
-    for vault_namespace in cfg.getList(['vault', 'namespaces'], ['']):
+    max_retries = 3
+    error_occured = False
 
-        logger.info("Check vault_namespace: %s", vault_namespace)
-        try:
-            providers = cfg.getList(['vault.namespace.{}'.format(vault_namespace), 'providers'],
-                                    ['ec2', 'azure', 'gce'])
-
-            if 'azure' in providers:
-                instances = Azure(vault_namespace).list_resource_groups()
-                instances = [azure_to_local_instance(i, vault_namespace) for i in instances]
-                logger.info("Got %d resources groups from Azure", len(instances))
-                sync_csp_to_local_db(instances, ProviderChoice.AZURE, vault_namespace)
-
-            if 'ec2' in providers:
-                instances = []
-                for region in cfg.getList(['ec2', 'regions'], EC2(vault_namespace).list_regions()):
-                    instances_csp = EC2(vault_namespace).list_instances(region=region)
-                    instances += [ec2_to_local_instance(i, vault_namespace, region) for i in instances_csp]
-                    logger.info("Got %d instances from EC2 in region %s", len(instances), region)
-                sync_csp_to_local_db(instances, ProviderChoice.EC2, vault_namespace)
-
-            if 'gce' in providers:
-                instances = GCE(vault_namespace).list_all_instances()
-                instances = [gce_to_local_instance(i, vault_namespace) for i in instances]
-                logger.info("Got %d instances from GCE", len(instances))
-                sync_csp_to_local_db(instances, ProviderChoice.GCE, vault_namespace)
-
-        except Exception as e:
-            logger.exception("Update failed!")
-            send_mail(type(e).__name__ + ' on Update', traceback.format_exc())
+    for vault_namespace in cfg.getList(['vault', 'namespaces'], ['']):
+        for provider in cfg.getList(['vault.namespace.{}'.format(vault_namespace), 'providers'],
+                                    ['ec2', 'azure', 'gce']):
+            logger.info("Check provider %s in vault_namespace %s", provider, vault_namespace)
+            email_text = set()
+            for n in range(max_retries):
+                try:
+                    _update_provider(provider, vault_namespace)
+                except Exception:
+                    logger.exception("Update failed for {} in namespace {}".format(provider, vault_namespace))
+                    email_text.add(traceback.format_exc())
+                    time.sleep(5)
+                else:
+                    break
+            else:
```

Comment:
didn't get where is `if` for this `else`

Reply:
... or `except`

Reply:
it's the for :) lovely python...
https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

Reply:
ah I see :)